src/modules/cpu/kernel/transpose_generic.hpp

```diff
@@ -0,0 +1,441 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "rppdefs.h"
+#include "rpp_cpu_simd.hpp"
+#include "rpp_cpu_common.hpp"
+using namespace std;
+
+void compute_strides(Rpp32u *strides, Rpp32u *shape, Rpp32u nDim)
+{
+    if (nDim > 0)
+    {
+        uint64_t v = 1;
+        for (int i = nDim - 1; i > 0; i--)
+        {
+            strides[i] = v;
+            v *= shape[i];
+        }
+        strides[0] = v;
+    }
+}
+
+void increment_ndim_ptr(Rpp32f **dstPtr, Rpp32u nDim, Rpp32u increment)
+{
+    for(int i = 0; i < nDim; i++)
+        dstPtr[i] += increment;
+}
+
+void rpp_store16_f32_f32_channelwise(Rpp32f **dstPtr, __m128 *p)
+{
+    _mm_storeu_ps(dstPtr[0], p[0]);
+    _mm_storeu_ps(dstPtr[1], p[1]);
+    _mm_storeu_ps(dstPtr[2], p[2]);
+    _mm_storeu_ps(dstPtr[3], p[3]);
+}
+
+void compute_2d_transpose(Rpp32f *srcPtrTemp, Rpp32f *dstPtrTemp, Rpp32u height, Rpp32u width, Rpp32u srcRowStride, Rpp32u dstRowStride)
+{
+    Rpp32u alignedRows = (height / 4) * 4;
+    Rpp32u alignedCols = (width / 8) * 8;
+    Rpp32u vectorIncrement = 8;
+    Rpp32u dstRowVectorStride = vectorIncrement * dstRowStride;
+
+    int i = 0;
+    for(; i < alignedRows; i += 4)
+    {
+        Rpp32s k = (i / 4);
+        Rpp32f *srcPtrRow[4], *dstPtrRow[8];
+
+        for(int j = 0; j < 4; j++)
+            srcPtrRow[j] = srcPtrTemp + (i + j) * srcRowStride;
+        for(int j = 0; j < 8; j++)
+            dstPtrRow[j] = dstPtrTemp + j * dstRowStride + k * 4;
+
+        Rpp32u vectorLoopCount = 0;
+        for(; vectorLoopCount < alignedCols; vectorLoopCount += vectorIncrement)
+        {
+            __m256 pSrc[4];
+            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow[0], &pSrc[0]);
+            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow[1], &pSrc[1]);
+            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow[2], &pSrc[2]);
+            rpp_simd_load(rpp_load8_f32_to_f32_avx, srcPtrRow[3], &pSrc[3]);
+
+            __m128 pDst[8];
+            compute_transpose4x8_avx(pSrc, pDst);
+            rpp_simd_store(rpp_store4_f32_to_f32, dstPtrRow[0], &pDst[0]);
+            rpp_simd_store(rpp_store4_f32_to_f32, dstPtrRow[1], &pDst[1]);
```

Comment:
__m256* pointer casting back to AVX before the store or other method might cut number of stores in half? Pls check

Reply:
Here we are storing 4 values in 8 different pointers
So combing 2 SSE to one AVX register might not work, since in that case i can only store in 4 pointers instead of 8

we can discuss this over a call if needed. Please let me know